go-controller/hybrid-overlay/pkg/controller/node_linux.go

```diff
@@ -29,56 +31,231 @@ const (
+	ip4 := podIP.To4()
+	if ip4 == nil {
+		return ""
+	}
+	return fmt.Sprintf("%02x%02x%02x%02x", ip4[0], ip4[1], ip4[2], ip4[3])
+}
+
+func (n *NodeController) waitForNamespace(name string) (*kapi.Namespace, error) {
+	var namespaceBackoff = wait.Backoff{Duration: 1 * time.Second, Steps: 7, Factor: 1.5, Jitter: 0.1}
+	var namespace *kapi.Namespace
+	if err := wait.ExponentialBackoff(namespaceBackoff, func() (bool, error) {
+		var err error
+		namespace, err = n.wf.GetNamespace(name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// Namespace not found; retry
+				return false, nil
+			}
+			klog.Warningf("error getting namespace: %v", err)
+			return false, err
+		}
+		return true, nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to get namespace object: %v", err)
+	}
+	return namespace, nil
 }
 
-func (n *NodeController) addOrUpdatePod(pod *kapi.Pod) error {
+func (n *NodeController) addOrUpdatePod(pod *kapi.Pod, ignoreLearn bool) error {
 	podIPs, podMAC, err := getPodDetails(pod, n.nodeName)
 	if err != nil {
 		klog.V(5).Infof("cleaning up hybrid overlay pod %s/%s because %v", pod.Namespace, pod.Name, err)
 		return n.deletePod(pod)
 	}
 
-	cookie := podToCookie(pod)
-	for _, podIP := range podIPs {
-		_, _, err = util.RunOVSOfctl("add-flow", extBridgeName,
-			fmt.Sprintf("table=10,cookie=0x%s,priority=100,ip,nw_dst=%s,actions=set_field:%s->eth_src,set_field:%s->eth_dst,output:ext", cookie, podIP.IP, n.drMAC, podMAC))
+	namespace, err := n.waitForNamespace(pod.Namespace)
+	if err != nil {
+		return err
+	}
+	namespaceExternalGw := namespace.GetAnnotations()[hotypes.HybridOverlayExternalGw]
+	namespaceVTEP := namespace.GetAnnotations()[hotypes.HybridOverlayVTEP]
+	if !n.initialized {
+		node, err := n.wf.GetNode(n.nodeName)
 		if err != nil {
-			return fmt.Errorf("failed to add flows for pod %s/%s: %v", pod.Namespace, pod.Name, err)
+			return fmt.Errorf("hybrid overlay not initialized on %s, and failed to get node data: %v",
+				n.nodeName, err)
+		}
+		if err = n.ensureHybridOverlayBridge(node); err != nil {
+			return fmt.Errorf("failed to ensure hybrid overlay in pod handler: %v", err)
 		}
 	}
+	if n.drMAC == nil || n.drIP == nil {
+		return fmt.Errorf("empty values for DR MAC: %s or DR IP: %s on node %s", n.drMAC, n.drIP, n.nodeName)
+	}
+	var flows []string
+	for _, podIP := range podIPs {
+		cookie := podIPToCookie(podIP.IP)
+		if cookie == "" {
+			continue
+		}
+		// table 10 is pod dispatch - Incoming vxlan traffic towards pods
+		flows = append(flows, fmt.Sprintf(
+			"table=10,cookie=0x%s,priority=100,ip,nw_dst=%s,"+
+				"actions=set_field:%s->eth_src,set_field:%s->eth_dst,output:ext",
+			cookie, podIP.IP, n.drMAC.String(), podMAC))
+
+		if namespaceExternalGw == "" || namespaceVTEP == "" {
+			klog.Infof("Skipping hybrid overlay gw mode setup for pod %s, namespace does not have hybrid"+
+				"annotations, external gw: %s, VTEP: %s", pod.Name, namespaceExternalGw, namespaceVTEP)
+			break
```

Comment:
@dcbw I think this needs to be a continue so we handle the rest of the IPs for non-exgw (windows) hybrid overlay case

Reply:
we also need to program flows here for this case so we dont miss programming the windows flows